File: src/panda_controller/panda_gripper.cpp
// Custom
#include "panda_controller/panda_gripper.hpp"



// NAMESPACE ==================================================================
/// @brief Namespace of panda_errors.
namespace panda_controller {



// CLASSES ====================================================================
PandaGripper::PandaGripper(const bool REAL_ROBOT) {
    // Initialize
    _real_robot = REAL_ROBOT;
    _current_width = -1;  // use homing for override it
    _timeout = 0;         // inf

    if (REAL_ROBOT) {
        try {
            // Create action client for homing
            _homing_client_ptr.reset(
                new GripperHomingClient("franka_gripper/homing", true));

            // Create action client for move
            _move_client_ptr.reset(
                new GripperMoveClient("franka_gripper/move", true));

            // Create action client for grasp
            _grasp_client_ptr.reset(
                new GripperGraspClient("franka_gripper/grasp", true));

        } catch (const std::runtime_error &err) {
            throw PandaGripperErr(
                "PandaGripper()",
                "Impossible create action clients for franka_gripper");
        }

    } else {
        // Init MoveGroupInterface with hand
        std::string move_group_name = "hand";
        try {
            _hand_ptr.reset(new moveit::planning_interface::MoveGroupInterface(
                move_group_name));

        } catch (const std::runtime_error &err) {
            throw PandaGripperErr(
                "PandaGripper()",
                "Impossible initialize MoveGroupInterface with '" +
                    move_group_name + "'");
        }
    }
}


double PandaGripper::_normalize(const double VAL, const double MIN,
                                const double MAX) {
    if (VAL < MIN)
        return MIN;
    if (VAL > MAX)
        return MAX;
    return VAL;
}


void PandaGripper::_moveHand(const double WIDTH_NORMALIZED) {
    // Prepare fingers
    std::vector<double> fingers;
    fingers.push_back(WIDTH_NORMALIZED / 2.0);
    fingers.push_back(WIDTH_NORMALIZED / 2.0);
    _hand_ptr->setJointValueTarget(fingers);

    // Perform movement
    auto res = _hand_ptr->move();

    // Errors check
    if (res != moveit::planning_interface::MoveItErrorCode::SUCCESS)
        throw PandaGripperErr("_moveHand()", "move()", "Failure");
}


void PandaGripper::setTiming(const double TIMEOUT) {
    _timeout = TIMEOUT;
}


double PandaGripper::getWidth() {
    return _current_width;
}


void PandaGripper::homing() {
    if (_real_robot) {
        // Create homing goal
        auto goal = franka_gripper::HomingGoal();
        _homing_client_ptr->sendGoal(goal);

        // Wait for result
        if (!_homing_client_ptr->waitForResult(ros::Duration(_timeout)))
            throw PandaGripperErr("homing()", "waitForResult()", "Timeout");
    }

    // Update current width
    _current_width = OPEN;
}


void PandaGripper::move(const double WIDTH, const double SPEED) {
    double width_normalized =
        _normalize(WIDTH, panda_gripper::MIN_WIDTH, panda_gripper::MAX_WIDTH);

    if (_real_robot) {
        // Create move goal
        auto goal = franka_gripper::MoveGoal();
        goal.width = width_normalized;
        goal.speed = SPEED;
        _move_client_ptr->sendGoal(goal);

        // Wait for result
        if (!_move_client_ptr->waitForResult(ros::Duration(_timeout)))
            throw PandaGripperErr("move()", "waitForResult()", "Timeout");

    } else
        _moveHand(width_normalized);

    // Update current width
    _current_width = width_normalized;
}


void PandaGripper::grasp(const double WIDTH, const double SPEED,
                         const double FORCE, const double EPSILON_INNER,
                         const double EPSILON_OUTER) {
    double width_normalized =
        _normalize(WIDTH, panda_gripper::MIN_WIDTH, panda_gripper::MAX_WIDTH);
    double force_normalized =
        _normalize(WIDTH, panda_gripper::MIN_FORCE, panda_gripper::MAX_FORCE);

    if (_real_robot) {
        // Create move goal
        auto goal = franka_gripper::GraspGoal();
        goal.width = width_normalized;
        goal.speed = SPEED;
        goal.force = force_normalized;
        goal.epsilon.inner = EPSILON_INNER;
        goal.epsilon.outer = EPSILON_OUTER;
        _grasp_client_ptr->sendGoal(goal);

        // Wait for result
        if (!_grasp_client_ptr->waitForResult(ros::Duration(_timeout)))
            throw PandaGripperErr("grasp()", "waitForResult()", "Timeout");

    } else
        _moveHand(width_normalized);

    // Update current width
    _current_width = width_normalized;
}

}  // namespace panda_controller

File: scripts/src/panda_gripper.py
#!/usr/bin/env python3

# ROS and Moveit
from moveit_commander.exception import MoveItCommanderException
import moveit_commander
import actionlib
import rospy

# Action msg
from franka_gripper.msg import HomingAction, MoveAction, GraspAction, StopAction
from franka_gripper.msg import HomingGoal, MoveGoal, GraspGoal, StopGoal, GraspEpsilon

# Other
import sys, os
sys.path.append(os.path.abspath(os.path.dirname(__file__)))



class PandaGripper():
    OPEN  = 0.08  # [m]

    def __init__(self, delay = 0, real_robot = False):
        # Attributes
        self._MIN_WIDTH = 0.0                  # [m] closed
        self._MAX_WIDTH = PandaGripper.OPEN    # [m] opened
        self._MIN_FORCE = 0.01                 # [N]
        self._MAX_FORCE = 50.0                 # [N]
        self._timeout = 0.0                    # [sec] -> inf
        self._current_width = -1               # [m] use homing for override it
        self._real_robot = real_robot

        if self._real_robot:
            # Create action clients
            self._client_homing = actionlib.SimpleActionClient('franka_gripper/homing', HomingAction)
            self._client_move = actionlib.SimpleActionClient('franka_gripper/move', MoveAction)
            self._client_grasp = actionlib.SimpleActionClient('franka_gripper/grasp', GraspAction)
            self._client_stop = actionlib.SimpleActionClient('franka_gripper/stop', StopAction)

            # Wait action servers
            self._client_homing.wait_for_server(rospy.Duration.from_sec(self._timeout))
            self._client_move.wait_for_server(rospy.Duration.from_sec(self._timeout))
            self._client_grasp.wait_for_server(rospy.Duration.from_sec(self._timeout))
            self._client_stop.wait_for_server(rospy.Duration.from_sec(self._timeout))

        else:
            # enable hand
            self.hand = moveit_commander.MoveGroupCommander("hand")

        # wait for correct loading
        rospy.sleep(delay)


    def _normalize(self, val, min, max):
        if val < min:
            print("Adust min")
            return min
        if val > max:
            print("Adjust max")
            return max
        return val 


    # HAND---------------------------------------------------------------------
    def _moveHand(self, width_normalized):
        """[width] is the distance between fingers (min=0.0, max=0.08)"""
        try:
            self.hand.set_joint_value_target([width_normalized/2.0, width_normalized/2.0])
            self.hand.plan()
            self.hand.go(wait=True)
        except MoveItCommanderException:
            return False
        return True
      
    
    def getGripperWidth(self):
        return self._current_width


    def homingGripper(self):
        if self._real_robot:
            # Create goal
            goal = HomingGoal()
            # Send goal
            self._client_homing.send_goal_and_wait(goal, rospy.Duration.from_sec(self._timeout))
        else:
            self._moveHand(PandaGripper.OPEN)
        self._current_width = PandaGripper.OPEN
      

    def moveGripper(self, width, speed=0.5):
        width_normalized = self._normalize(width, self._MIN_WIDTH, self._MAX_WIDTH)
        if self._real_robot:
            # Create goal
            goal = MoveGoal()
            goal.width = self.safe_width(width_normalized)
            goal.speed = speed
            # Send goal
            self._client_move.send_goal_and_wait(goal, rospy.Duration.from_sec(self._timeout))
        else:
            self._moveHand(width_normalized)
        self._current_width = width_normalized
      

    def graspGripper(self, width, speed=0.5, force=10, epsilon_inner=0.02, epsilon_outer=0.02):
        width_normalized = self._normalize(width, self._MIN_WIDTH, self._MAX_WIDTH)
        force_normalized = self._normalize(force, self._MIN_FORCE, self._MAX_FORCE)
        if self._real_robot:
            # Create goal
            goal = GraspGoal()
            goal.width = width_normalized
            goal.epsilon = GraspEpsilon(epsilon_inner, epsilon_outer)
            goal.speed = speed
            goal.force = force_normalized
            # Send goal
            self._client_grasp.send_goal_and_wait(goal, rospy.Duration.from_sec(self._timeout))
        else:
            self._moveHand(width_normalized)
            print("GRASPING...")
        self._current_width = width_normalized
      
      
    def stopGripper(self):
        if self._real_robot:
            self.gripper.stop()
        else:
            print("hand.stopGripper() NOT IMPLEMENTED YET")



def test_gripper(gripper):

    print("getGripperWidth(): {}\n".format(gripper.getGripperWidth()))

    print("homingGripper()")
    gripper.homingGripper()
    rospy.sleep(1)
    print("Current gripper width: {}\n".format(gripper.getGripperWidth()))

    print("moveGripper(width=10, speed=0.01)")
    gripper.moveGripper(width=10, speed=0.01)
    rospy.sleep(1)
    print("Current gripper width: {}\n".format(gripper.getGripperWidth()))

    print("graspGripper(width=0.02, epsilon_inner=0.02, epsilon_outer=0.02, speed=0.01, force=10)")
    gripper.graspGripper(width=0.02, epsilon_inner=0.02, epsilon_outer=0.02, speed=0.01, force=10)
    rospy.sleep(1)
    print("Current gripper width: {}\n".format(gripper.getGripperWidth()))



if __name__ == '__main__':
    """TEST"""
    import sys
    # Initialize moveit_commander and rospy
    moveit_commander.roscpp_initialize(sys.argv)
    rospy.init_node('panda_gripper_moveit_test_node', anonymous=True)

    try:
        # Inizialize movegroupinterface
        gripper = PandaGripper(delay=1, real_robot=False)

        test_gripper(gripper)

    except rospy.ROSInterruptException:
        print("ROS interrupted")
      